Instalaciones e importaciones

In [1]:
import pandas as pd 
import numpy as np
import tools 
import warnings
warnings.filterwarnings("ignore")

Extraemos los datos de la pestaña "Hechos" del archivo de excel homicidios y observamos las primeras filas. 

In [2]:
hom_hechos = pd.read_excel("homicidios.xlsx", sheet_name="HECHOS")
hom_hechos.head(10)

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00   0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00   5   
5  2016-0008           1 2016-01-24  2016   1  24  18:30:00  18   
6  2016-0009           1 2016-01-24  2016   1  24  19:10:00  19   
7  2016-0010           1 2016-01-29  2016   1  29  15:20:00  15   
8  2016-0012           1 2016-02-08  2016   2   8  01:20:00   1   
9  2016-0013           1 2016-02-10  2016   2  10  11:30:00  11   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                         AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   
5             AV 27 DE FEBRERO Y AV ESCALADA       AVENIDA  ...     NaN   
6               NOGOYA Y JOAQUIN V. GONZALES         CALLE  ...     NaN   
7        AV GENERAL PAZ Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
8         AV BELGRANO Y BERNARDO DE IRIGOYEN       AVENIDA  ...     NaN   
9                         AV ENTRE RIOS 1366       AVENIDA  ...  1366.0   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   
3             VILLEGAS, CONRADO, GRAL.   
4              SAENZ PE?A, LUIS, PRES.   
5                         ESCALADA AV.   
6                 GONZALEZ, JOAQUIN V.   
7                  DE LOS CORRALES AV.   
8                IRIGOYEN, BERNARDO DE   
9                                  NaN   

                               Dirección Normalizada COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
5                   27 DE FEBRERO AV. y ESCALADA AV.      8   
6                      NOGOYA y GONZALEZ, JOAQUIN V.     11   
7               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
8               BELGRANO AV. e IRIGOYEN, BERNARDO DE      1   
9                                ENTRE RIOS AV. 1366      1   

                                 XY (CABA)         pos x         pos y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
5   Point (101721.59002217 93844.25656649)  -58.44451316  -34.68475866   
6   Point (96545.87592078 102330.67262199)  -58.50095869  -34.60825440   
7    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
8  Point (107595.35084333 101797.50052813)  -58.38048577  -34.61303893   
9  Point (106616.41069662 100496.44662323)  -58.39114932  -34.62477387   

      PARTICIPANTES VICTIMA      ACUSADO  
0         MOTO-AUTO    MOTO         AUTO  
1    AUTO-PASAJEROS    AUTO    PASAJEROS  
2         MOTO-AUTO    MOTO         AUTO  
3           MOTO-SD    MOTO           SD  
4    MOTO-PASAJEROS    MOTO    PASAJEROS  
5  MOTO-OBJETO FIJO    MOTO  OBJETO FIJO  
6         MOTO-AUTO    MOTO         AUTO  
7         MOTO-AUTO    MOTO         AUTO  
8       MOTO-CARGAS    MOTO       CARGAS  
9       PEATON-AUTO  PEATON         AUTO  




#### Observación general

El conjunto de datos contiene 21 variables(columnas) y 696 registros que veremos a continuación:

In [3]:
# Columnas del dataset
hom_hechos.columns


Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Altura', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO'],
      dtype='object')

In [4]:
# Cantidad de registros
print("La cantidad de registros es:", hom_hechos.shape[0])

La cantidad de registros es: 696


Vamos a estandarizar los nombres de los nombres de las columnas.

In [5]:
#La primera letra con mayúscula
hom_hechos.columns = [x.capitalize() for x in hom_hechos.columns]
#Cambiamos los guiones por espacios
hom_hechos.columns = hom_hechos.columns.str.replace("_", " ")

hom_hechos.columns

Index(['Id', 'N victimas', 'Fecha', 'Aaaa', 'Mm', 'Dd', 'Hora', 'Hh',
       'Lugar del hecho', 'Tipo de calle', 'Calle', 'Altura', 'Cruce',
       'Dirección normalizada', 'Comuna', 'Xy (caba)', 'Pos x', 'Pos y',
       'Participantes', 'Victima', 'Acusado'],
      dtype='object')

In [6]:
#Renombramos algunas columnas
hom_hechos = hom_hechos.rename(columns={'N victimas': 'Cantidad víctimas',
                                        'Aaaa':'Año',
                                        'Mm':'Mes',
                                        'Dd':'Día',
                                        'Hh':'Hora entera',
                                        'Xy (caba)':'XY (CABA)',
                                        'Victima': 'Víctima'})

hom_hechos.columns

Index(['Id', 'Cantidad víctimas', 'Fecha', 'Año', 'Mes', 'Día', 'Hora',
       'Hora entera', 'Lugar del hecho', 'Tipo de calle', 'Calle', 'Altura',
       'Cruce', 'Dirección normalizada', 'Comuna', 'XY (CABA)', 'Pos x',
       'Pos y', 'Participantes', 'Víctima', 'Acusado'],
      dtype='object')

#### Datos nulos
En relación a los datos faltantes, se identifican 747 valores nulos, lo que equivale al 5.1% del total de datos, distribuidos en las columnas "Altura", "Cruce" y "Dirección normalizada".

En particular, la columna "Altura" presenta una ausencia del 81.5% de los datos, ya que este campo se completa únicamente cuando el incidente ocurre en algún punto específico de la calle (altura de la calle). Dado que la mayoría de los incidentes registrados ocurrieron en esquinas, se considera que este dato no es relevante para el análisis y se decide eliminarlo del conjunto de datos. Adicionalmente, esta información también está presente en las columnas "Lugar del hecho" y "Dirección normalizada". Se proporciona un ejemplo de este comportamiento, donde las dos primeras filas corresponden a esquinas y la tercera a un punto entre la cuadra.

In [7]:
hom_hechos[["Altura", "Lugar del hecho", "Dirección normalizada"]].head()

Altura                            Lugar del hecho  \
0     NaN  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ   
1     NaN          AV GRAL PAZ  Y AV DE LOS CORRALES   
2  2034.0                         AV ENTRE RIOS 2034   
3     NaN      AV LARRAZABAL Y GRAL VILLEGAS CONRADO   
4     NaN  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA    

                               Dirección normalizada  
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...  
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.  
2                                ENTRE RIOS AV. 2034  
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.  
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.

In [8]:
#Eliminamos la columna "Altura"
hom_hechos = hom_hechos.drop("Altura", axis=1)
hom_hechos.columns

Index(['Id', 'Cantidad víctimas', 'Fecha', 'Año', 'Mes', 'Día', 'Hora',
       'Hora entera', 'Lugar del hecho', 'Tipo de calle', 'Calle', 'Cruce',
       'Dirección normalizada', 'Comuna', 'XY (CABA)', 'Pos x', 'Pos y',
       'Participantes', 'Víctima', 'Acusado'],
      dtype='object')

En lo que respecta a la columna "Cruce", se trata de una variable categórica de tipo string, que abarca 317 valores únicos y presenta un 24.6% de valores faltantes. A diferencia de la variable anterior, la ausencia de datos en esta columna se debe a que no se registra información cuando el incidente ocurre en un cruce. Esta variable podría resultar relevante para analizar si los incidentes tienen lugar en esquinas o en otros puntos de la cuadra. A continuación, se examina la columna "Cruce" en relación con "Lugar del hecho" y "Tipo de calle"

In [9]:
# veremos los valores nulos de "cruce" en relación con "Lugar del hecho" y "Tipo de calle"
hom_hechos[hom_hechos["Cruce"].isnull()][["Cruce", "Lugar del hecho", "Tipo de calle"]].head(10)

Cruce                                    Lugar del hecho Tipo de calle
2    NaN                                 AV ENTRE RIOS 2034       AVENIDA
9    NaN                                 AV ENTRE RIOS 1366       AVENIDA
14   NaN                                       SUIPACHA 156         CALLE
33   NaN                                          LIMA 1483         CALLE
35   NaN  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....     AUTOPISTA
38   NaN                         AUTOPISTA LUGONES PK 10000     AUTOPISTA
42   NaN                            AV LUIS MARIA CAMPOS 30       AVENIDA
43   NaN                                          LIMA 1471         CALLE
50   NaN                           MAGARIÑOS CERVANTES 3900         CALLE
59   NaN                                   AV LA PLATA 2384       AVENIDA

In [10]:
# Veremos los valores NO nulos 
hom_hechos[hom_hechos["Cruce"].notnull()][["Cruce", "Lugar del hecho", "Tipo de calle"]].head(20)

Cruce  \
0   FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                   DE LOS CORRALES AV.   
3              VILLEGAS, CONRADO, GRAL.   
4               SAENZ PE?A, LUIS, PRES.   
5                          ESCALADA AV.   
6                  GONZALEZ, JOAQUIN V.   
7                   DE LOS CORRALES AV.   
8                 IRIGOYEN, BERNARDO DE   
10                                 VERA   
11             QUINQUELA MARTIN, BENITO   
12                     JANER, ANA MARIA   
13    GARCIA GRANDE DE ZEQUEIRA, SEVERO   
15                          CURAPALIGUE   
16                       ALSINA, ADOLFO   
17                    27 DE FEBRERO AV.   
18                                VEDIA   
19                        DIAZ, AVELINO   
20                       9 DE JULIO AV.   
21                               DEHEZA   
22                      ROCA, CNEL. AV.   

                                      Lugar del hecho Tipo de calle  
0           AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  
1                   AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  
3               AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  
4           AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  
5                      AV 27 DE FEBRERO Y AV ESCALADA       AVENIDA  
6                        NOGOYA Y JOAQUIN V. GONZALES         CALLE  
7                 AV GENERAL PAZ Y AV DE LOS CORRALES      GRAL PAZ  
8                  AV BELGRANO Y BERNARDO DE IRIGOYEN       AVENIDA  
10                         AV SCALABRINI ORTIZ Y VERA       AVENIDA  
11  AV REGIMIENTO DE PATRICIOS Y QUINQUELA MARTIN ...       AVENIDA  
12                   MARIANO ACOSTA Y ANA MARIA JANER       AVENIDA  
13  LISANDRO DE LA TORRE Y SEVERO GARCIA GRANDE DE...         CALLE  
15                        AV DIRECTORIO Y CURAPALIGUE       AVENIDA  
16                      AV 9 DE JULIO Y ADOLFO ALSINA       AVENIDA  
17                     AV GRAL PAZ Y AV 27 DE FEBRERO      GRAL PAZ  
18                                   HOLMBERG Y VEDIA         CALLE  
19                         CURAPALIGUE Y AVELINO DIAZ       AVENIDA  
20                           VIAMONTE Y AV 9 DE JULIO       AVENIDA  
21                                AV CABILDO Y DEHEZA       AVENIDA  
22                         AV GRAL PAZ Y AV CNEL ROCA      GRAL PAZ

Es evidente que los valores faltantes corresponden a incidentes que ocurrieron en algún punto de la calle que no constituye una esquina o intersección de dos calles. Por consiguiente, se opta por actualizar esta columna asignando el valor "sí" a los casos que involucran hechos en cruces de calles y "no" a aquellos que no se ubican en una intersección de calles.

In [11]:
# Reemplazamos los valores no nulos por "no" y los nulos por "si"
hom_hechos["Cruce"] = np.where(hom_hechos["Cruce"].notnull(), "SI", "NO")
hom_hechos[["Cruce", "Lugar del hecho", "Tipo de calle"]]

Cruce                            Lugar del hecho Tipo de calle
0      SI  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA
1      SI          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ
2      NO                         AV ENTRE RIOS 2034       AVENIDA
3      SI      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA
4      SI  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA
..    ...                                        ...           ...
691    SI                          AV. RIESTRA Y MOM       AVENIDA
692    SI                    AU DELLEPIANE Y LACARRA     AUTOPISTA
693    SI                        AV. GAONA Y TERRADA       AVENIDA
694    NO                         AV. EVA PERON 4071       AVENIDA
695    NO                    PADRE CARLOS MUJICA 709         CALLE

[696 rows x 3 columns]

La tercera columna que presenta valores faltantes es "Dirección normalizada", y contiene únicamente 8 datos nulos, equivalente al 1.1% de los datos en dicha columna. A continuación, se examinan detalladamente esos registros.

In [12]:
hom_hechos[hom_hechos["Dirección normalizada"].isnull()][["Dirección normalizada", "Lugar del hecho", "Tipo de calle"]].head(20)

Dirección normalizada                          Lugar del hecho  \
38                    NaN               AUTOPISTA LUGONES PK 10000   
106                   NaN         AU BUENOS AIRES - LA PLATA KM. 4   
119                   NaN                                       SD   
180                   NaN  AU PERITO MORENO Y RAMAL ENLACE AU1/AU6   
181                   NaN                       AU DELLEPIANE 2400   
313                   NaN                 AUTOPISTA LUGONES KM 4.7   
546                   NaN             LUGONES, LEOPOLDO AV. KM 6,1   
621                   NaN         AU BUENOS AIRES LA PLATA  KM 4,5   

    Tipo de calle  
38      AUTOPISTA  
106     AUTOPISTA  
119         CALLE  
180     AUTOPISTA  
181     AUTOPISTA  
313     AUTOPISTA  
546     AUTOPISTA  
621     AUTOPISTA

Aunque no se cuenta con información que permita llenar estos datos faltantes con valores específicos, se ha optado por no eliminar los registros. En su lugar, se decide imputar a los valores nulos con la etiqueta 'SD', indicando que se trata de "Sin Dato"

In [13]:
#Rellenamos nulos con SD 
hom_hechos["Dirección normalizada"].fillna("SD", inplace=True)
#Verificamos los datos 
hom_hechos[hom_hechos["Dirección normalizada"]=="SD"][['Dirección normalizada', 'Lugar del hecho', 'Tipo de calle']]


Dirección normalizada                          Lugar del hecho  \
38                     SD               AUTOPISTA LUGONES PK 10000   
106                    SD         AU BUENOS AIRES - LA PLATA KM. 4   
119                    SD                                       SD   
180                    SD  AU PERITO MORENO Y RAMAL ENLACE AU1/AU6   
181                    SD                       AU DELLEPIANE 2400   
313                    SD                 AUTOPISTA LUGONES KM 4.7   
546                    SD             LUGONES, LEOPOLDO AV. KM 6,1   
621                    SD         AU BUENOS AIRES LA PLATA  KM 4,5   

    Tipo de calle  
38      AUTOPISTA  
106     AUTOPISTA  
119         CALLE  
180     AUTOPISTA  
181     AUTOPISTA  
313     AUTOPISTA  
546     AUTOPISTA  
621     AUTOPISTA

#### Datos duplicados

No se detectaron filas duplicadas en el conjunto de datos. La variable "Id" presenta 696 valores distintos.

In [14]:
print(f"La columna 'Id' cuenta con {len(hom_hechos['Id'].unique())} valores únicos.")
tools.ver_duplicados(hom_hechos, 'Id')

La columna 'Id' cuenta con 696 valores únicos.


'No hay duplicados'

#### Variables

El informe presenta diversos tipos de variables, y a continuación se proporciona la clasificación de tipo de dato para cada columna.

In [15]:
#Tipo de dato por columna
tools.ver_variables(hom_hechos)

nombre_campo                                         tipo_datos
0                      Id                                    [<class 'str'>]
1       Cantidad víctimas                                    [<class 'int'>]
2                   Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...
3                     Año                                    [<class 'int'>]
4                     Mes                                    [<class 'int'>]
5                     Día                                    [<class 'int'>]
6                    Hora  [<class 'datetime.time'>, <class 'str'>, <clas...
7             Hora entera                     [<class 'int'>, <class 'str'>]
8         Lugar del hecho                                    [<class 'str'>]
9           Tipo de calle                                    [<class 'str'>]
10                  Calle                   [<class 'str'>, <class 'float'>]
11                  Cruce                                    [<class 'str'>]
12  Dirección normalizada                                    [<class 'str'>]
13                 Comuna                                    [<class 'int'>]
14              XY (CABA)                                    [<class 'str'>]
15                  Pos x                                    [<class 'str'>]
16                  Pos y                                    [<class 'str'>]
17          Participantes                                    [<class 'str'>]
18                Víctima                                    [<class 'str'>]
19                Acusado                                    [<class 'str'>]

Las columnas "Hora", "Hora entera" y "Calle" muestran diversos tipos de datos dentro de una misma columna, lo que requiere una revisión individual para asignar un único tipo de dato a cada una.

En primer lugar, se examina la columna "Hora".

In [16]:
hom_hechos["Hora"].apply(type).value_counts()

Hora
<class 'datetime.time'>        608
<class 'str'>                   85
<class 'datetime.datetime'>      3
Name: count, dtype: int64

Se decide convertir todos los datos a datetime.time porque la mayoría de datos son de ese tipo.

In [17]:
# Cambiamos el tipo de dato
hom_hechos["Hora"] = hom_hechos["Hora"].apply(lambda x: tools.convertir_a_time(x))
# Verificamos la cantidad de valores por tipo de dato en la columna "hora"
print("Tipos de datos:")
print(hom_hechos["Hora"].apply(type).value_counts())
print("Registro(s) con NoneType:")
hom_hechos[hom_hechos["Hora"].isna()]


Tipos de datos:
Hora
<class 'datetime.time'>    695
<class 'NoneType'>           1
Name: count, dtype: int64
Registro(s) con NoneType:


Id  Cantidad víctimas      Fecha   Año  Mes  Día  Hora  \
518  2019-0103                  1 2019-12-18  2019   12   18  None   

    Hora entera          Lugar del hecho Tipo de calle           Calle Cruce  \
518          SD  PAZ, GRAL. AV. Y GRIVEO      GRAL PAZ  PAZ, GRAL. AV.    SI   

       Dirección normalizada  Comuna                               XY (CABA)  \
518  PAZ, GRAL. AV. y GRIVEO      11  Point (94643.11254058 103831.57115061)   

            Pos x         Pos y Participantes Víctima Acusado  
518  -58.52169422  -34.59471640     MOTO-MOTO    MOTO    MOTO

Es evidente que la información sobre la hora está ausente tanto en las columnas 'Hora' como en 'Hora entera'. Para abordarlo, se ha tomado la decisión de imputar un valor.

In [18]:
# Calculamos la hora mas común
hora_moda = hom_hechos["Hora"].mode().iloc[0]
print(f"La hora más común es: {hora_moda}")
# Reemplazamos el valor None por la hora más común
hom_hechos["Hora"].fillna(hora_moda, inplace=True)
# Verificamos el tipo de dato para la columna
hom_hechos["Hora"].apply(type).value_counts()

La hora más común es: 09:00:00


Hora
<class 'datetime.time'>    696
Name: count, dtype: int64

Como se mencionó anteriormente, la columna 'Hora entera' muestra dos tipos de datos debido a la presencia de 'SD' en el mismo incidente analizado previamente. En consecuencia, se ha decidido imputar la hora calculada más común a este dato faltante

In [19]:
# Verificamos el tipo de dato
hom_hechos["Hora entera"].apply(type).value_counts()

Hora entera
<class 'int'>    695
<class 'str'>      1
Name: count, dtype: int64

In [20]:
# Imputamos la hora moda al dato faltante
hom_hechos["Hora entera"] = hom_hechos["Hora entera"].apply(lambda x: int(hora_moda.hour) if x == "SD" else x)
# Se verifica el tipo de dato
hom_hechos["Hora entera"].apply(type).value_counts()

Hora entera
<class 'int'>    696
Name: count, dtype: int64

In [21]:
# Verificamos el registro completado
hom_hechos[hom_hechos["Id"]=="2019-0103"]

Id  Cantidad víctimas      Fecha   Año  Mes  Día      Hora  \
518  2019-0103                  1 2019-12-18  2019   12   18  09:00:00   

     Hora entera          Lugar del hecho Tipo de calle           Calle Cruce  \
518            9  PAZ, GRAL. AV. Y GRIVEO      GRAL PAZ  PAZ, GRAL. AV.    SI   

       Dirección normalizada  Comuna                               XY (CABA)  \
518  PAZ, GRAL. AV. y GRIVEO      11  Point (94643.11254058 103831.57115061)   

            Pos x         Pos y Participantes Víctima Acusado  
518  -58.52169422  -34.59471640     MOTO-MOTO    MOTO    MOTO

Por último, la columna "Calle" muestra dos tipos de datos. A pesar de que la información de la variable indica que se trata de una columna de cadenas de texto (strings), se han identificado algunos valores de tipo float. Se procede a examinar los registros que contienen este último tipo de dato.

In [22]:
# Verificamos el tipo de dato
print("Tipos de datos:")
print(hom_hechos["Hora entera"].apply(type).value_counts())
# Observamos el registro con tipo de dato float
print("Registro con tipo float:")
hom_hechos[hom_hechos["Calle"].apply(lambda x: isinstance(x, float))]

Tipos de datos:
Hora entera
<class 'int'>    696
Name: count, dtype: int64
Registro con tipo float:


Id  Cantidad víctimas      Fecha   Año  Mes  Día      Hora  \
119  2016-0151                  1 2016-11-18  2016   11   18  20:35:00   

     Hora entera Lugar del hecho Tipo de calle Calle Cruce  \
119           20              SD         CALLE   NaN    NO   

    Dirección normalizada  Comuna    XY (CABA) Pos x Pos y Participantes  \
119                    SD       0  Point (. .)     .     .     PEATON-SD   

    Víctima Acusado  
119  PEATON      SD

Se puede observar la ausencia del dato correspondiente a la calle donde tuvo lugar el incidente. En consecuencia, se ha tomado la decisión de completar este valor faltante imputando "SD".

In [23]:
# Imputamos SD 
hom_hechos["Calle"].fillna("SD", inplace=True)
hom_hechos[hom_hechos["Id"]=="2016-0151"]

Id  Cantidad víctimas      Fecha   Año  Mes  Día      Hora  \
119  2016-0151                  1 2016-11-18  2016   11   18  20:35:00   

     Hora entera Lugar del hecho Tipo de calle Calle Cruce  \
119           20              SD         CALLE    SD    NO   

    Dirección normalizada  Comuna    XY (CABA) Pos x Pos y Participantes  \
119                    SD       0  Point (. .)     .     .     PEATON-SD   

    Víctima Acusado  
119  PEATON      SD

Verificamos nuevamente los tipos de variables.

In [24]:
tools.ver_variables(hom_hechos)

nombre_campo                                         tipo_datos
0                      Id                                    [<class 'str'>]
1       Cantidad víctimas                                    [<class 'int'>]
2                   Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...
3                     Año                                    [<class 'int'>]
4                     Mes                                    [<class 'int'>]
5                     Día                                    [<class 'int'>]
6                    Hora                          [<class 'datetime.time'>]
7             Hora entera                                    [<class 'int'>]
8         Lugar del hecho                                    [<class 'str'>]
9           Tipo de calle                                    [<class 'str'>]
10                  Calle                                    [<class 'str'>]
11                  Cruce                                    [<class 'str'>]
12  Dirección normalizada                                    [<class 'str'>]
13                 Comuna                                    [<class 'int'>]
14              XY (CABA)                                    [<class 'str'>]
15                  Pos x                                    [<class 'str'>]
16                  Pos y                                    [<class 'str'>]
17          Participantes                                    [<class 'str'>]
18                Víctima                                    [<class 'str'>]
19                Acusado                                    [<class 'str'>]

#### Categorías Víctima y Acusado

In [25]:

hom_hechos["Víctima"].unique()

array(['MOTO', 'AUTO', 'PEATON', 'SD', 'CARGAS', 'BICICLETA', 'PASAJEROS',
       'MOVIL', 'OBJETO FIJO', 'PEATON_MOTO'], dtype=object)

In [26]:
hom_hechos["Acusado"].unique()

array(['AUTO', 'PASAJEROS', 'SD', 'OBJETO FIJO', 'CARGAS', 'MOTO',
       'MULTIPLE', 'OTRO', 'BICICLETA', 'TREN'], dtype=object)

No se muestran particularidades en la categoría "Acusado". Sin embargo, al revisar las categorías relacionadas con "Víctima", se identifican dos clasificaciones, a saber, "Objeto fijo" y "Peatón moto", que no están incluidas en el diccionario de datos. Es necesario examinar estos registros para efectuar las correcciones correspondientes.

In [27]:
hom_hechos[hom_hechos["Víctima"].isin(["OBJETO FIJO", "PEATON_MOTO"])]

Id  Cantidad víctimas      Fecha   Año  Mes  Día      Hora  \
230  2017-0108                  2 2017-09-02  2017    9    2  04:53:08   
583  2020-0063                  2 2020-12-05  2020   12    5  07:10:00   

     Hora entera           Lugar del hecho Tipo de calle           Calle  \
230            4   AV. GRAL. PAZ Y MACHAIN      GRAL PAZ  PAZ, GRAL. AV.   
583            7  NUEVA YORK Y ALTA GRACIA         CALLE      NUEVA YORK   

    Cruce     Dirección normalizada  Comuna  \
230    SI  PAZ, GRAL. AV. y MACHAIN      12   
583    SI  NUEVA YORK y ALTA GRACIA      11   

                                  XY (CABA)         Pos x         Pos y  \
230  Point (97098.48468623 109019.96106626)  -58.49491054  -34.54795581   
583  Point (94080.62190808 102083.62453795)  -58.52783814  -34.61047001   

        Participantes      Víctima Acusado  
230  AUTO-OBJETO FIJO  OBJETO FIJO    AUTO  
583  PEATON_MOTO-MOTO  PEATON_MOTO    MOTO

Se muestran dos registros, cada uno asociado a un término no registrado en el diccionario de datos. Se tomó la decisión de sustituír dichos términos por "OTRO", ya que no se dispone de otra información que permita indicar el vehículo de la víctima.

In [28]:
# Cambiamos por OTRO y verificamos
hom_hechos["Víctima"] = hom_hechos["Víctima"].replace({"OBJETO FIJO":"OTRO", "PEATON_MOTO":"OTRO"})
hom_hechos[hom_hechos["Id"].isin(["2017-0108", "2020-0063"])]


Id  Cantidad víctimas      Fecha   Año  Mes  Día      Hora  \
230  2017-0108                  2 2017-09-02  2017    9    2  04:53:08   
583  2020-0063                  2 2020-12-05  2020   12    5  07:10:00   

     Hora entera           Lugar del hecho Tipo de calle           Calle  \
230            4   AV. GRAL. PAZ Y MACHAIN      GRAL PAZ  PAZ, GRAL. AV.   
583            7  NUEVA YORK Y ALTA GRACIA         CALLE      NUEVA YORK   

    Cruce     Dirección normalizada  Comuna  \
230    SI  PAZ, GRAL. AV. y MACHAIN      12   
583    SI  NUEVA YORK y ALTA GRACIA      11   

                                  XY (CABA)         Pos x         Pos y  \
230  Point (97098.48468623 109019.96106626)  -58.49491054  -34.54795581   
583  Point (94080.62190808 102083.62453795)  -58.52783814  -34.61047001   

        Participantes Víctima Acusado  
230  AUTO-OBJETO FIJO    OTRO    AUTO  
583  PEATON_MOTO-MOTO    OTRO    MOTO

#### Variables de ubicación

En el análisis previo, se evidenció la ausencia de datos geográficos en las columnas "XY_(CABA)", "Pos_x" y "Pos_y". Se procede a examinar los registros correspondientes a dicha carencia.


In [29]:
hom_hechos[hom_hechos["XY (CABA)"]=="Point (. .)"]

Id  Cantidad víctimas      Fecha   Año  Mes  Día      Hora  \
35   2016-0049                  1 2016-04-17  2016    4   17  00:00:00   
38   2016-0052                  1 2016-04-20  2016    4   20  20:00:00   
71   2016-0096                  1 2016-07-25  2016    7   25  07:00:00   
106  2016-0136                  1 2016-10-25  2016   10   25  00:00:00   
119  2016-0151                  1 2016-11-18  2016   11   18  20:35:00   
139  2016-0174                  1 2016-12-27  2016   12   27  00:00:00   
176  2017-0042                  1 2017-04-10  2017    4   10  09:00:00   
180  2017-0050                  2 2017-04-28  2017    4   28  11:08:08   
181  2017-0051                  1 2017-05-01  2017    5    1  03:47:47   
256  2017-0140                  1 2017-11-19  2017   11   19  23:22:17   
313  2018-0039                  1 2018-04-21  2018    4   21  22:15:00   
546  2020-0026                  1 2020-05-17  2020    5   17  06:40:00   
559  2020-0039                  1 2020-09-01  2020    9    1  19:17:42   
621  2021-0023                  1 2021-03-01  2021    3    1  09:20:00   

     Hora entera                                    Lugar del hecho  \
35             0  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....   
38            20                         AUTOPISTA LUGONES PK 10000   
71             7       AUTOPISTA DELLEPIANE LUIS TTE. GRAL. KM. 2,3   
106            0                   AU BUENOS AIRES - LA PLATA KM. 4   
119           20                                                 SD   
139            0                               AUTOPISTA 25 DE MAYO   
176            9                      AV. LEOPOLDO LUGONES PKM 6900   
180           11            AU PERITO MORENO Y RAMAL ENLACE AU1/AU6   
181            3                                 AU DELLEPIANE 2400   
256           23                        AU ARTURO FRONDIZI PKM 3100   
313           22                           AUTOPISTA LUGONES KM 4.7   
546            6                       LUGONES, LEOPOLDO AV. KM 6,1   
559           19                                    MURGUIONDO 2700   
621            9                   AU BUENOS AIRES LA PLATA  KM 4,5   

    Tipo de calle                                       Calle Cruce  \
35      AUTOPISTA  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI    NO   
38      AUTOPISTA                       LUGONES, LEOPOLDO AV.    NO   
71      AUTOPISTA        AUTOPISTA DELLEPIANE LUIS TTE. GRAL.    NO   
106     AUTOPISTA           AUTOPISTA BUENOS AIRES - LA PLATA    NO   
119         CALLE                                          SD    NO   
139     AUTOPISTA                        AUTOPISTA 25 DE MAYO    NO   
176      GRAL PAZ                       LUGONES, LEOPOLDO AV.    NO   
180     AUTOPISTA                     AUTOPISTA PERITO MORENO    NO   
181     AUTOPISTA        AUTOPISTA DELLEPIANE LUIS TTE. GRAL.    NO   
256     AUTOPISTA  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI    NO   
313     AUTOPISTA                       LUGONES, LEOPOLDO AV.    NO   
546     AUTOPISTA                       LUGONES, LEOPOLDO AV.    NO   
559         CALLE                                  MURGUIONDO    NO   
621     AUTOPISTA           AUTOPISTA BUENOS AIRES - LA PLATA    NO   

                          Dirección normalizada  Comuna    XY (CABA)  \
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI       4  Point (. .)   
38                                           SD      13  Point (. .)   
71         AUTOPISTA DELLEPIANE LUIS TTE. GRAL.       8  Point (. .)   
106                                          SD       4  Point (. .)   
119                                          SD       0  Point (. .)   
139                        AUTOPISTA 25 DE MAYO       0  Point (. .)   
176                       LUGONES, LEOPOLDO AV.      14  Point (. .)   
180                                          SD       9  Point (. .)   
181                                          SD       7  Point (. .)   
256  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI  

In [30]:
hom_hechos["Pos x"] = hom_hechos["Pos x"].replace(".", 0)
hom_hechos["Pos y"] = hom_hechos["Pos y"].replace(".", 0)
hom_hechos["XY (CABA)"] = hom_hechos["XY (CABA)"].replace("Point (. .)", 0)


# Verificamos que los valores reemplazados
hom_hechos[hom_hechos["Id"]=="2020-0026"]

Id  Cantidad víctimas      Fecha   Año  Mes  Día      Hora  \
546  2020-0026                  1 2020-05-17  2020    5   17  06:40:00   

     Hora entera               Lugar del hecho Tipo de calle  \
546            6  LUGONES, LEOPOLDO AV. KM 6,1     AUTOPISTA   

                     Calle Cruce Dirección normalizada  Comuna XY (CABA)  \
546  LUGONES, LEOPOLDO AV.    NO                    SD      14         0   

    Pos x Pos y     Participantes Víctima      Acusado  
546     0     0  MOTO-OBJETO FIJO    MOTO  OBJETO FIJO

Ahora extraemos los datos de la pestaña "VICTIMAS" del archivo de excel homicidios y observamos las primeras filas. 

In [31]:
hom_victimas = pd.read_excel("homicidios.xlsx", sheet_name= "VICTIMAS")
hom_victimas

ID_hecho      FECHA  AAAA  MM  DD                   ROL VICTIMA  \
0    2016-0001 2016-01-01  2016   1   1             CONDUCTOR    MOTO   
1    2016-0002 2016-01-02  2016   1   2             CONDUCTOR    AUTO   
2    2016-0003 2016-01-03  2016   1   3             CONDUCTOR    MOTO   
3    2016-0004 2016-01-10  2016   1  10             CONDUCTOR    MOTO   
4    2016-0005 2016-01-21  2016   1  21             CONDUCTOR    MOTO   
..         ...        ...   ...  ..  ..                   ...     ...   
712  2021-0092 2021-12-12  2021  12  12                PEATON  PEATON   
713  2021-0093 2021-12-13  2021  12  13  PASAJERO_ACOMPAÑANTE    MOTO   
714  2021-0094 2021-12-20  2021  12  20  PASAJERO_ACOMPAÑANTE    MOTO   
715  2021-0095 2021-12-30  2021  12  30             CONDUCTOR    MOTO   
716  2021-0096 2021-12-15  2021  12  15             CONDUCTOR    AUTO   

          SEXO EDAD  FECHA_FALLECIMIENTO  
0    MASCULINO   19  2016-01-01 00:00:00  
1    MASCULINO   70  2016-01-02 00:00:00  
2    MASCULINO   30  2016-01-03 00:00:00  
3    MASCULINO   18                   SD  
4    MASCULINO   29  2016-02-01 00:00:00  
..         ...  ...                  ...  
712   FEMENINO   50  2021-12-12 00:00:00  
713   FEMENINO   18  2021-12-18 00:00:00  
714   FEMENINO   43  2021-12-20 00:00:00  
715  MASCULINO   27  2022-01-02 00:00:00  
716  MASCULINO   60  2021-12-20 00:00:00  

[717 rows x 10 columns]

#### Observación general

El conjunto de datos presenta 10 variables(columnas) y 717 registros(filas).

In [32]:
#Columnas del dataset
hom_victimas.columns

Index(['ID_hecho', 'FECHA', 'AAAA', 'MM', 'DD', 'ROL', 'VICTIMA', 'SEXO',
       'EDAD', 'FECHA_FALLECIMIENTO'],
      dtype='object')

In [33]:
# cantidad de registros
x = hom_victimas.shape[0]
print("La cantidad de registros es:", x)

La cantidad de registros es: 717


Como se vio en el dataset anterior, necesitamos estandarizar el nombre de las columnas. 

In [34]:
# Colocamos la primera letra en mayúscula
hom_victimas.columns = [x.capitalize() for x in hom_victimas.columns]
# Reemplazamos los guiones por espacios
hom_victimas.columns = hom_victimas.columns.str.replace("_", " ")
# Renombramos algunas columnas
hom_victimas = hom_victimas.rename(columns={"Id hecho": "Id",
                                                      "Aaaa":"Año",
                                                      "Mm":"Mes",
                                                      "Dd":"Día",
                                                      "Victima":"Víctima"})
hom_victimas.columns

Index(['Id', 'Fecha', 'Año', 'Mes', 'Día', 'Rol', 'Víctima', 'Sexo', 'Edad',
       'Fecha fallecimiento'],
      dtype='object')

#### Datos nulos

Se verifica si el dataset contiene datos nulos

In [35]:
hom_victimas.isna().any()

Id                     False
Fecha                  False
Año                    False
Mes                    False
Día                    False
Rol                    False
Víctima                False
Sexo                   False
Edad                   False
Fecha fallecimiento    False
dtype: bool

#### Datos duplicados

Se verifica si el dataset contiene duplicados.

In [36]:
hom_victimas[hom_victimas.duplicated()]

Empty DataFrame
Columns: [Id, Fecha, Año, Mes, Día, Rol, Víctima, Sexo, Edad, Fecha fallecimiento]
Index: []

Como podemos ver, no hay filas duplicadas. 
Ahora, se verifica si el id es único o no. 

In [37]:
tools.ver_duplicados(hom_victimas, "Id")

Id      Fecha   Año  Mes  Día                   Rol Víctima  \
29   2016-0041 2016-03-29  2016    3   29             CONDUCTOR    MOTO   
30   2016-0041 2016-03-29  2016    3   29  PASAJERO_ACOMPAÑANTE    MOTO   
98   2016-0126 2016-09-18  2016    9   18             CONDUCTOR    AUTO   
99   2016-0126 2016-09-18  2016    9   18  PASAJERO_ACOMPAÑANTE    AUTO   
163  2017-0026 2017-02-26  2017    2   26  PASAJERO_ACOMPAÑANTE    AUTO   
164  2017-0026 2017-02-26  2017    2   26             CONDUCTOR    AUTO   
173  2017-0035 2017-03-23  2017    3   23             CONDUCTOR    AUTO   
174  2017-0035 2017-03-23  2017    3   23  PASAJERO_ACOMPAÑANTE    AUTO   
175  2017-0035 2017-03-23  2017    3   23  PASAJERO_ACOMPAÑANTE    AUTO   
177  2017-0036 2017-03-29  2017    3   29             CONDUCTOR    MOTO   
176  2017-0036 2017-03-29  2017    3   29  PASAJERO_ACOMPAÑANTE    MOTO   
186  2017-0050 2017-04-28  2017    4   28             CONDUCTOR    MOTO   
187  2017-0050 2017-04-28  2017    4   28  PASAJERO_ACOMPAÑANTE    MOTO   
237  2017-0108 2017-09-02  2017    9    2  PASAJERO_ACOMPAÑANTE    AUTO   
238  2017-0108 2017-09-02  2017    9    2  PASAJERO_ACOMPAÑANTE    AUTO   
242  2017-0112 2017-09-10  2017    9   10  PASAJERO_ACOMPAÑANTE    AUTO   
243  2017-0112 2017-09-10  2017    9   10  PASAJERO_ACOMPAÑANTE    AUTO   
253  2017-0126 2017-10-14  2017   10   14  PASAJERO_ACOMPAÑANTE    AUTO   
254  2017-0126 2017-10-14  2017   10   14             CONDUCTOR    AUTO   
278  2017-0153 2017-12-11  2017   12   11             CONDUCTOR    MOTO   
277  2017-0153 2017-12-11  2017   12   11  PASAJERO_ACOMPAÑANTE    MOTO   
300  2018-0015 2018-02-06  2018    2    6             CONDUCTOR    MOTO   
301  2018-0015 2018-02-06  2018    2    6  PASAJERO_ACOMPAÑANTE    MOTO   
312  2018-0026 2018-03-09  2018    3    9             CONDUCTOR   MOVIL   
313  2018-0026 2018-03-09  2018    3    9  PASAJERO_ACOMPAÑANTE   MOVIL   
325  2018-0038 2018-04-20  2018    4   20                PEATON  PEATON   
326  2018-0038 2018-04-20  2018    4   20                PEATON  PEATON   
373  2018-0085 2018-08-03  2018    8    3  PASAJERO_ACOMPAÑANTE    MOTO   
374  2018-0085 2018-08-03  2018    8    3             CONDUCTOR    MOTO   
388  2018-0098 2018-09-09  2018    9    9  PASAJERO_ACOMPAÑANTE    AUTO   
387  2018-0098 2018-09-09  2018    9    9             CONDUCTOR    AUTO   
418  2018-0128 2018-11-25  2018   11   25  PASAJERO_ACOMPAÑANTE    AUTO   
419  2018-0128 2018-11-25  2018   11   25  PASAJERO_ACOMPAÑANTE    AUTO   
444  2019-0010 2019-02-02  2019    2    2                PEATON  PEATON   
445  2019-0010 2019-02-02  2019    2    2                PEATON  PEATON   
576  2020-0038 2020-08-30  2020    8   30             CONDUCTOR    AUTO   
577  2020-0038 2020-08-30  2020    8   30             CONDUCTOR    AUTO   
602  2020-0063 2020-12-05  2020   12    5                PEATON  PEATON   
603  2020-0063 2020-12-05  2020   12    5  PASAJERO_ACOMPAÑANTE    MOTO   
616  2020-0077 2020-12-25  2020   12   25             CONDUCTOR    MOTO   
617  2020-0077 2020-12-25  2020   12   25  PASAJERO_ACOMPAÑANTE    MOTO   

          Sexo Edad  Fecha fallecimiento  
29   MASCULINO   54  2016-03-29 00:00:00  
30   MASCULINO   SD  2016-03-30 00:00:00  
98   MASCULINO   37                   SD  
99   MASCULINO   60                   SD  
163   FEMENINO   23  2017-02-26 00:00:00  
164  MASCULINO   19  2017-02-26 00:00:00  
173  MASCULINO   28  2017-03-23 00:00:00  
174  MASCULINO   32  2017-03-23 00:00:00  
175  MASCULINO   30  2017-03-23 00:00:00  
177  MASCULINO   20  2017-03-29 00:00:00  
176   FEMENINO   SD  2017-03-29 00:00:00  
186  MASCULINO   46  2017-04-28 00:00:00  
187  MASCULINO   16  2017-04-28 00:00:00  
237  MASCULINO   17  2017-09-02 00:00:00  
238  MASCULINO   10  2017-09-02 00:00:00  
242  MASCULINO    1                   SD  
243  MASCULINO   SD  2017-09-13 00:00:00  
253  MASCULINO   39                   SD  
254  MASCULINO   36  2017-10-14 00:00:00  
2

Se identifican duplicados en los identificadores ("Id"), sin embargo, tras un análisis se determina que estos se corresponden con un mismo incidente que involucra a varias víctimas distintas. Por este motivo, se opta por mantener dichos registros sin alteraciones.



#### Tipos de variables

Se verifica los distintos tipos de variables contenidos en el dataset.

In [38]:
tools.ver_variables(hom_victimas)

nombre_campo                                         tipo_datos
0                   Id                                    [<class 'str'>]
1                Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...
2                  Año                                    [<class 'int'>]
3                  Mes                                    [<class 'int'>]
4                  Día                                    [<class 'int'>]
5                  Rol                                    [<class 'str'>]
6              Víctima                                    [<class 'str'>]
7                 Sexo                                    [<class 'str'>]
8                 Edad                     [<class 'int'>, <class 'str'>]
9  Fecha fallecimiento       [<class 'datetime.datetime'>, <class 'str'>]

En las columnas "Edad" y "Fecha fallecimiento" se detecta la presencia de múltiples tipos de datos, por lo que es necesario revisar cada uno de ellos y asignar un único formato para homogeneizar la información.

Se analiza en primer lugar la columna "Edad"

In [39]:
# Cantidad de valores por tipo de dato en la columna 'edad'
tipos_datos = hom_victimas["Edad"].apply(type).value_counts()
print("Los tipos de datos son:")
print(tipos_datos)
print(f"Los datos {tipos_datos.index[1]} representan el {round((tipos_datos[1]/tipos_datos.sum()),2)}")


Los tipos de datos son:
Edad
<class 'int'>    664
<class 'str'>     53
Name: count, dtype: int64
Los datos <class 'str'> representan el 0.07


Se identifican 53 registros en la columna, los cuales contienen datos en formato de cadena, representando aproximadamente el 7% del total de registros en dicha columna. A continuación, se procede a examinar detalladamente estos registros.

In [40]:
hom_victimas[hom_victimas["Edad"].apply(lambda x: isinstance(x, str))]


Id      Fecha   Año  Mes  Día                   Rol    Víctima  \
30   2016-0041 2016-03-29  2016    3   29  PASAJERO_ACOMPAÑANTE       MOTO   
33   2016-0045 2016-04-11  2016    4   11             CONDUCTOR       MOTO   
35   2016-0048 2016-04-15  2016    4   15                PEATON     PEATON   
36   2016-0049 2016-04-17  2016    4   17                    SD         SD   
39   2016-0052 2016-04-20  2016    4   20                    SD       MOTO   
55   2016-0077 2016-06-13  2016    6   13                PEATON     PEATON   
63   2016-0085 2016-06-29  2016    6   29                    SD       MOTO   
72   2016-0096 2016-07-25  2016    7   25             CONDUCTOR       MOTO   
89   2016-0115 2016-09-02  2016    9    2                    SD         SD   
93   2016-0119 2016-09-04  2016    9    4  PASAJERO_ACOMPAÑANTE         SD   
108  2016-0136 2016-10-25  2016   10   25             CONDUCTOR       MOTO   
119  2016-0148 2016-11-16  2016   11   16             CONDUCTOR       MOTO   
121  2016-0151 2016-11-18  2016   11   18                PEATON     PEATON   
138  2016-0171 2016-12-25  2016   12   25             CONDUCTOR       MOTO   
141  2016-0174 2016-12-27  2016   12   27                    SD         SD   
142  2016-0175 2016-12-27  2016   12   27                PEATON     PEATON   
165  2017-0027 2017-02-26  2017    2   26                PEATON     PEATON   
176  2017-0036 2017-03-29  2017    3   29  PASAJERO_ACOMPAÑANTE       MOTO   
182  2017-0042 2017-04-10  2017    4   10             CONDUCTOR       MOTO   
225  2017-0093 2017-07-25  2017    7   25                PEATON     PEATON   
243  2017-0112 2017-09-10  2017    9   10  PASAJERO_ACOMPAÑANTE       AUTO   
275  2017-0151 2017-12-04  2017   12    4                PEATON     PEATON   
277  2017-0153 2017-12-11  2017   12   11  PASAJERO_ACOMPAÑANTE       MOTO   
299  2018-0014 2018-01-31  2018    1   31                PEATON     PEATON   
328  2018-0040 2018-04-27  2018    4   27                PEATON     PEATON   
359  2018-0071 2018-07-10  2018    7   10             CONDUCTOR       MOTO   
367  2018-0079 2018-07-23  2018    7   23                PEATON     PEATON   
371  2018-0083 2018-08-02  2018    8    2             CONDUCTOR       MOTO   
377  2018-0088 2018-08-17  2018    8   17             CONDUCTOR       MOTO   
391  2018-0101 2018-09-21  2018    9   21             CONDUCTOR       AUTO   
421  2018-0130 2018-12-03  2018   12    3             CONDUCTOR       MOTO   
444  2019-0010 2019-02-02  2019    2    2                PEATON     PEATON   
446  2019-0011 2019-02-10  2019    2   10                PEATON     PEATON   
454  2019-0019 2019-03-06  2019    3    6             CONDUCTOR       MOTO   
457  2019-0022 2019-03-15  2019    3   15  PASAJERO_ACOMPAÑANTE       AUTO   
461  2019-0026 2019-03-28  2019    3   28                PEATON     PEATON   
463  2019-0028 2019-04-06  2019    4    6                PEATON     PEATON   
469  2019-0034 2019-04-21  2019    4   21             CONDUCTOR       MOTO   
479  2019-0043 2019-05-21  2019    5   21              CICLISTA  BICICLETA   
498  2019-0063 2019-08-04  2019    8    4  PASAJERO_ACOMPAÑANTE       MOTO   
506  2019-0071 2019-08-27  2019    8   27                PEATON     PEATON   
529  2019-0094 2019-11-18  2019   11   18  PASAJERO_ACOMPAÑANTE       MOTO   
530  2019-0095 2019-11-20  2019   11   20             CONDUCTOR       MOTO   
548  2020-0010 2020-02-08  2020    2    8             CONDUCTOR       MOTO   
552  2020-0014 2020-02-27  2020    2   27             CONDUCTOR       MOTO   
555  2020-0017 2020-03-05  2020    3    5                PEATON     PEATON   
563  2020-0025 2020-05-12  2020    5   12                PEATON     PEATON   
572  2020-0034 2020-08-02  2020    8    2                PEATON     PEATON   
579  2020-0040 2020-09-12  2020    9   12             CONDUCTOR       AUTO   
631  2021-0012 2021-01-29  2021    1   29                PEATON     PEATON   
639  2021-0020 2021-02-21  2021   

Con el objetivo de llenar los datos faltantes en la columna "Edad", se planea imputar dichos valores utilizando el promedio de las edades de las víctimas, considerando el factor 'Sexo'. Se opta por primero completar estos valores con el valor más frecuente entre las víctimas.

In [41]:
# Imputamos valor más frecuente en la columna sexo
tools.imputa_valor_frecuente(hom_victimas, "Sexo")

El valor mas frecuente es: MASCULINO


Ahora se imputa la edad, teniendo en cuenta el sexo.

In [42]:
tools.imputa_edad_media_segun_sexo(hom_victimas)

La edad promedio de Femenino es 51 y de Masculino es 40


Se corroboran los datos imputados

In [43]:
id_ejemplos = ["2016-0049","2016-0151","2016-0171"]
hom_victimas[hom_victimas["Id"].isin(id_ejemplos)]  

Id      Fecha   Año  Mes  Día        Rol Víctima       Sexo  Edad  \
36   2016-0049 2016-04-17  2016    4   17         SD      SD  MASCULINO    39   
121  2016-0151 2016-11-18  2016   11   18     PEATON  PEATON  MASCULINO    39   
138  2016-0171 2016-12-25  2016   12   25  CONDUCTOR    MOTO  MASCULINO    39   

    Fecha fallecimiento  
36                   SD  
121                  SD  
138                  SD

En el primer ejemplo, se observa que "Rol" y "Víctima" también tienen faltantes de datos. Se revisa la cantidad de SD que hay en cada columna. 

In [44]:
print(f"La cantidad de SD en 'Rol' es de {len(hom_victimas[hom_victimas['Rol']=='SD'])}")
print(f"La cantidad de SD en 'Víctima' es de {len(hom_victimas[hom_victimas['Víctima']=='SD'])}")

La cantidad de SD en 'Rol' es de 11
La cantidad de SD en 'Víctima' es de 9


Dado que se constata que los datos faltantes son pocos, se toma la decisión de imputar el valor más frecuente para cada columna.

In [45]:
# Imputamos valor mas frecuente en la columna rol
tools.imputa_valor_frecuente(hom_victimas, "Rol")

El valor mas frecuente es: CONDUCTOR


In [46]:
# Imputamos valor mas frecuente en la columna victima
tools.imputa_valor_frecuente(hom_victimas, "Víctima")

El valor mas frecuente es: MOTO


Se revisan los mismos ejemplos anteriores, para ver estas últimas imputaciones.

In [47]:
hom_victimas[hom_victimas["Id"].isin(id_ejemplos)]

Id      Fecha   Año  Mes  Día        Rol Víctima       Sexo  Edad  \
36   2016-0049 2016-04-17  2016    4   17  CONDUCTOR    MOTO  MASCULINO    39   
121  2016-0151 2016-11-18  2016   11   18     PEATON  PEATON  MASCULINO    39   
138  2016-0171 2016-12-25  2016   12   25  CONDUCTOR    MOTO  MASCULINO    39   

    Fecha fallecimiento  
36                   SD  
121                  SD  
138                  SD

Finalmente, la columna "Fecha fallecimiento" muestra aparentemente  registros incompletos. Dado que la meta es analizar los datos en busca de patrones que orienten decisiones para reducir accidentes mortales, la información sobre la fecha de fallecimiento no aporta al análisis. Por tanto, se determina eliminar dicha columna.

In [48]:
hom_victimas = hom_victimas.drop("Fecha fallecimiento", axis = 1)
hom_victimas.head()

Id      Fecha   Año  Mes  Día        Rol Víctima       Sexo  Edad
0  2016-0001 2016-01-01  2016    1    1  CONDUCTOR    MOTO  MASCULINO    19
1  2016-0002 2016-01-02  2016    1    2  CONDUCTOR    AUTO  MASCULINO    70
2  2016-0003 2016-01-03  2016    1    3  CONDUCTOR    MOTO  MASCULINO    30
3  2016-0004 2016-01-10  2016    1   10  CONDUCTOR    MOTO  MASCULINO    18
4  2016-0005 2016-01-21  2016    1   21  CONDUCTOR    MOTO  MASCULINO    29

Nuevamente, se verifica los tipos de datos

In [49]:
tools.ver_variables(hom_victimas)

nombre_campo                                         tipo_datos
0           Id                                    [<class 'str'>]
1        Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...
2          Año                                    [<class 'int'>]
3          Mes                                    [<class 'int'>]
4          Día                                    [<class 'int'>]
5          Rol                                    [<class 'str'>]
6      Víctima                                    [<class 'str'>]
7         Sexo                                    [<class 'str'>]
8         Edad                                    [<class 'int'>]

En vista de la futura unión de este conjunto de datos con "hom_hechos", se plantea la posibilidad de eliminar las columnas 'Fecha', 'Año', 'Mes', 'Día' y 'Víctima', ya que contienen información duplicada presente en "hom_hechos". Lo verificaremos con un ejemplo.

In [50]:
hom_hechos[hom_hechos["Id"]=="2016-0005"]

Id  Cantidad víctimas      Fecha   Año  Mes  Día      Hora  \
4  2016-0005                  1 2016-01-21  2016    1   21  05:20:00   

   Hora entera                            Lugar del hecho Tipo de calle  \
4            5  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA   

          Calle Cruce                    Dirección normalizada  Comuna  \
4  SAN JUAN AV.    SI  SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.       1   

                                 XY (CABA)         Pos x         Pos y  \
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    Participantes Víctima    Acusado  
4  MOTO-PASAJEROS    MOTO  PASAJEROS

In [51]:
hom_victimas[hom_victimas["Id"]=="2016-0005"]

Id      Fecha   Año  Mes  Día        Rol Víctima       Sexo  Edad
4  2016-0005 2016-01-21  2016    1   21  CONDUCTOR    MOTO  MASCULINO    29

In [52]:
# Se eliminan las columnas repetidas
hom_victimas = hom_victimas.drop(["Fecha", "Año","Mes", "Día", "Víctima"], axis=1)
hom_victimas.columns

Index(['Id', 'Rol', 'Sexo', 'Edad'], dtype='object')

#### Combinar datos
Se unifican en un único dataframe los dos conjuntos de datos ingestados anteriormente.

In [53]:
df_homicidios = hom_victimas.merge(hom_hechos, on= "Id", how="left")
df_homicidios 

Id                   Rol       Sexo  Edad  Cantidad víctimas  \
0    2016-0001             CONDUCTOR  MASCULINO    19                  1   
1    2016-0002             CONDUCTOR  MASCULINO    70                  1   
2    2016-0003             CONDUCTOR  MASCULINO    30                  1   
3    2016-0004             CONDUCTOR  MASCULINO    18                  1   
4    2016-0005             CONDUCTOR  MASCULINO    29                  1   
..         ...                   ...        ...   ...                ...   
712  2021-0092                PEATON   FEMENINO    50                  1   
713  2021-0093  PASAJERO_ACOMPAÑANTE   FEMENINO    18                  1   
714  2021-0094  PASAJERO_ACOMPAÑANTE   FEMENINO    43                  1   
715  2021-0095             CONDUCTOR  MASCULINO    27                  1   
716  2021-0096             CONDUCTOR  MASCULINO    60                  1   

         Fecha   Año  Mes  Día      Hora  ...                         Calle  \
0   2016-01-01  2016    1    1  04:00:00  ...              PIEDRA BUENA AV.   
1   2016-01-02  2016    1    2  01:15:00  ...                PAZ, GRAL. AV.   
2   2016-01-03  2016    1    3  07:00:00  ...                ENTRE RIOS AV.   
3   2016-01-10  2016    1   10  00:00:00  ...                LARRAZABAL AV.   
4   2016-01-21  2016    1   21  05:20:00  ...                  SAN JUAN AV.   
..         ...   ...  ...  ...       ...  ...                           ...   
712 2021-12-12  2021   12   12  06:20:00  ...                 RIVADAVIA AV.   
713 2021-12-13  2021   12   13  17:10:00  ...                   RIESTRA AV.   
714 2021-12-20  2021   12   20  01:10:00  ...  DELLEPIANE, LUIS, TTE. GRAL.   
715 2021-12-30  2021   12   30  00:43:00  ...                     GAONA AV.   
716 2021-12-15  2021   12   15  10:30:00  ...                PERON, EVA AV.   

    Cruce                              Dirección normalizada Comuna  \
0      SI  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1      SI               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2      NO                                ENTRE RIOS AV. 2034      1   
3      SI          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4      SI            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
..    ...                                                ...    ...   
712    SI                     RIVADAVIA AV. y PUEYRREDON AV.      3   
713    SI                                  RIESTRA AV. y MOM      7   
714    SI         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.      9   
715    SI                                GAONA AV. y TERRADA     11   
716    NO                                PERON, EVA AV. 4071      9   

                                   XY (CABA)         Pos x         Pos y  \
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
712  Point (105258.35368554 102122.93231400)  -58.40596860  -34.61011987   
713   Point (102728.60090138 98186.24929177)  -58.43353773  -34.64561636   
714    Point (99624.29795829 97569.69801131)  -58.46739825  -34.65117757   
715   Point (99116.45492358 101045.23284826)  -58.47293407  -34.61984745   
716    Point (99324.54463985 97676.26932409)  -58.47066794  -34.65021673   

      Participantes Víctima    Acusado  
0         MOTO-AUTO    MOTO       AUTO  
1    AUTO-PASAJEROS    AUTO  PASAJEROS  
2         MOTO-AUTO    MOTO       AUTO  
3           MOTO-SD    MOTO         SD  
4    MOTO-PASAJEROS    MOTO  PASAJEROS  
..              ...     ...        ...  
712     PEATON-AUTO  PEATON       AUTO  
713       MOTO-AUTO    MOTO       AUTO  
714       MOTO-


Se verifican los tipos de datos y cantidad de nulos en este nuevo dataframe.

In [54]:
tools.ver_tipo_datos(df_homicidios)

nombre_campo                                         tipo_datos  \
0                      Id                                    [<class 'str'>]   
1                     Rol                                    [<class 'str'>]   
2                    Sexo                                    [<class 'str'>]   
3                    Edad                                    [<class 'int'>]   
4       Cantidad víctimas                                    [<class 'int'>]   
5                   Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
6                     Año                                    [<class 'int'>]   
7                     Mes                                    [<class 'int'>]   
8                     Día                                    [<class 'int'>]   
9                    Hora                          [<class 'datetime.time'>]   
10            Hora entera                                    [<class 'int'>]   
11        Lugar del hecho                                    [<class 'str'>]   
12          Tipo de calle                                    [<class 'str'>]   
13                  Calle                                    [<class 'str'>]   
14                  Cruce                                    [<class 'str'>]   
15  Dirección normalizada                                    [<class 'str'>]   
16                 Comuna                                    [<class 'int'>]   
17              XY (CABA)                     [<class 'str'>, <class 'int'>]   
18                  Pos x                     [<class 'str'>, <class 'int'>]   
19                  Pos y                     [<class 'str'>, <class 'int'>]   
20          Participantes                                    [<class 'str'>]   
21                Víctima                                    [<class 'str'>]   
22                Acusado                                    [<class 'str'>]   

    no_nulos_%  nulos_%  nulos  
0        100.0      0.0      0  
1        100.0      0.0      0  
2        100.0      0.0      0  
3        100.0      0.0      0  
4        100.0      0.0      0  
5        100.0      0.0      0  
6        100.0      0.0      0  
7        100.0      0.0      0  
8        100.0      0.0      0  
9        100.0      0.0      0  
10       100.0      0.0      0  
11       100.0      0.0      0  
12       100.0      0.0      0  
13       100.0      0.0      0  
14       100.0      0.0      0  
15       100.0      0.0      0  
16       100.0      0.0      0  
17       100.0      0.0      0  
18       100.0      0.0      0  
19       100.0      0.0      0  
20       100.0      0.0      0  
21       100.0      0.0      0  
22       100.0      0.0      0

Como se puede observar, no se presentan vzalores nulos y nohay diferentes tipos de datos. Por lo tanto, se toma la decisión de guardarlo para posteriormente analizarlo en la parte del Análisis Exploratorio de Datos. 

In [55]:
df_homicidios.to_csv("homicidios_cleaned.csv", index=False, encoding= "utf-8")
print(f"Se guardó el archivo")

Se guardó el archivo
